nexus/src/app/oximeter.rs

```diff
@@ -272,26 +281,48 @@ impl super::Nexus {
             )));
         }
 
-        Ok(if let Some(timeseries) = timeseries_list.into_iter().next() {
-            let next_start_time = end_time;
+        // If we received no data, exit early.
+        let timeseries =
+            if let Some(timeseries) = timeseries_list.into_iter().next() {
+                timeseries
+            } else {
+                return Ok(no_results());
+            };
 
-            dropshot::ResultsPage {
-                next_page: if next_start_time >= query.end_time {
-                    None
-                } else {
-                    Some(base64::encode_config(
-                        serde_json::to_string(&ResourceMetrics {
-                            start_time: next_start_time,
-                            end_time: query.end_time,
-                        })?,
-                        base64::URL_SAFE,
-                    ))
-                },
-                items: timeseries.measurements,
-            }
-        } else {
-            no_results()
-        })
+        // Otherwise, set up the next page to-be-queried before returning data.
+        //
+        // This is only relevant if we hit the limit of measurements when
+        // querying.
+        let next_page =
+            if timeseries.measurements.len() >= limit.get() as usize {
+                // The "next start time" is the latest value we saw from the query.
+                let next_start_time = timeseries
+                    .measurements
+                    .iter()
+                    .map(|m| m.timestamp())
+                    .reduce(|latest, timestamp| {
+                        if latest >= timestamp {
+                            latest
+                        } else {
+                            timestamp
+                        }
+                    })
+                    // Unwrap safety: This case shouldn't be possible to hit; we
+                    // only enter this conditional when we're at the (nonzero)
+                    // pagination limit of measurements.
+                    .expect("Expected measurements");
+                Some(base64::encode_config(
+                    serde_json::to_string(&ResourceMetrics {
+                        start_time: next_start_time,
+                        end_time: query.end_time,
+                    })?,
+                    base64::URL_SAFE,
+                ))
+            } else {
+                None
+            };
+
+        Ok(dropshot::ResultsPage { next_page, items: timeseries.measurements })
```

Comment:
FYI @iliana this - in addition to plumbing the `limit` variable down to the Clickhouse query - were what I mentioned changing in our meeting a couple minutes ago. 

Basically, I'm trying to make the caller unaware of the polling interval

Reply:
I meant to bring this up earlier, but we've been sort of conflating the polling interval with the data or measurement interval. The former describes the interval on which the `oximeter` collector will make a request to the producer. The latter describes the sampling interval of the underlying metric itself. Producers are free to generate as many samples as they want in a single request, so these are really independent.

You're right, I think, that this method here shouldn't really know or have to know about either one of them. Just wanted to make the distinction clear somewhere durable!